File: vm_native_retdec_gap_queue.py
#!/usr/bin/env python3
import argparse
import csv
from pathlib import Path


TRACE_DIR = Path("dumps/vmtail-wide-1m-w16")


FIELDS = [
    "rank",
    "offset",
    "end",
    "bytes",
    "semantic_covered_bytes",
    "semantic_uncovered_bytes",
    "coverage_x100",
    "kind",
    "name",
    "nbbs",
    "ninstrs",
    "outdegree",
    "priority_class",
    "next_action",
    "selected_range",
]


def read_tsv(path):
    with Path(path).open(newline="", errors="replace") as handle:
        return list(csv.DictReader(handle, delimiter="\t"))


def parse_int(value):
    return int(str(value), 0)


def semantic_ranges(path):
    ranges = []
    for row in read_tsv(path):
        if row.get("row_type") != "range":
            continue
        start = parse_int(row["start"])
        stop = parse_int(row["stop"])
        if stop > start:
            ranges.append((start, stop, row.get("source", "")))
    ranges.sort()
    return ranges


def overlap_bytes(start, stop, ranges):
    total = 0
    sources = set()
    for r_start, r_stop, source in ranges:
        if r_stop <= start:
            continue
        if r_start >= stop:
            break
        left = max(start, r_start)
        right = min(stop, r_stop)
        if right > left:
            total += right - left
            sources.add(source)
    return min(total, stop - start), sources


def classify(row, uncovered):
    size = parse_int(row["end"]) - parse_int(row["offset"])
    if uncovered == 0:
        return "covered_by_existing_semantic_c", "none"
    if row.get("kind") == "import_plt":
        return "import_plt_stub", "skip_import_thunk"
    if size <= 512:
        return "tier0_small_native_gap", "targeted_retdec_single_function"
    if size <= 2048:
        return "tier1_medium_native_gap", "targeted_retdec_single_function"
    if size <= 8192:
        return "tier2_large_native_gap", "targeted_retdec_with_timeout"
    return "tier3_huge_or_noisy_native_gap", "split_or_model_before_retdec"


def build_rows(inventory_path, coverage_path):
    ranges = semantic_ranges(coverage_path)
    rows = []
    for row in read_tsv(inventory_path):
        if row.get("section") != ".text":
            continue
        start = parse_int(row["offset"])
        stop = parse_int(row["end"])
        if stop <= start:
            continue
        size = stop - start
        covered, _sources = overlap_bytes(start, stop, ranges)
        uncovered = max(0, size - covered)
        if uncovered == 0:
            continue
        priority, action = classify(row, uncovered)
        rows.append({
            "rank": "0",
            "offset": row["offset"],
            "end": row["end"],
            "bytes": str(size),
            "semantic_covered_bytes": str(covered),
            "semantic_uncovered_bytes": str(uncovered),
            "coverage_x100": str((covered * 10000) // size if size else 0),
            "kind": row.get("kind", ""),
            "name": row.get("name", ""),
            "nbbs": row.get("nbbs", "0"),
            "ninstrs": row.get("ninstrs", "0"),
            "outdegree": row.get("outdegree", "0"),
            "priority_class": priority,
            "next_action": action,
            "selected_range": f"{row['offset']}-{row['end']}",
        })
    rows.sort(key=lambda r: (
        {"tier0_small_native_gap": 0, "tier1_medium_native_gap": 1, "tier2_large_native_gap": 2}.get(r["priority_class"], 3),
        -int(r["semantic_uncovered_bytes"]),
        int(r["offset"], 0),
    ))
    for i, row in enumerate(rows, 1):
        row["rank"] = str(i)
    return rows


def cstr(value):
    return '"' + str(value).replace("\\", "\\\\").replace('"', '\\"') + '"'


def emit_tsv(rows):
    writer = csv.DictWriter(__import__("sys").stdout, fieldnames=FIELDS, delimiter="\t", lineterminator="\n")
    writer.writeheader()
    writer.writerows(rows)


def emit_md(rows):
    classes = {}
    for row in rows:
        classes[row["priority_class"]] = classes.get(row["priority_class"], 0) + 1
    print("# Native RetDec Gap Queue\n")
    print("Ranks native function skeletons whose byte ranges are not yet covered by semantic RetDec/native C sidecars.\n")
    print(f"- queued functions: {len(rows)}")
    print("- class mix: " + ", ".join(f"{key}:{value}" for key, value in sorted(classes.items())) + "\n")
    print("| rank | range | bytes | uncovered | class | action | name |")
    print("| ---: | --- | ---: | ---: | --- | --- | --- |")
    for row in rows[:80]:
        print(
            f"| {row['rank']} | `{row['selected_range']}` | {row['bytes']} | "
            f"{row['semantic_uncovered_bytes']} | `{row['priority_class']}` | "
            f"`{row['next_action']}` | `{row['name']}` |"
        )


def emit_c(rows):
    print("/*")
    print(" * Native RetDec gap queue.")
    print(" * Ranks native function skeletons that still lack semantic C sidecar coverage.")
    print(" */")
    print("#include <stdint.h>")
    print("")
    print("typedef struct VMNativeRetdecGapQueueRow {")
    print("    uint32_t rank;")
    print("    uint64_t offset;")
    print("    uint64_t end;")
    print("    uint32_t bytes;")
    print("    uint32_t semantic_covered_bytes;")
    print("    uint32_t semantic_uncovered_bytes;")
    print("    uint32_t coverage_x100;")
    print("    const char *kind;")
    print("    const char *name;")
    print("    uint32_t nbbs;")
    print("    uint32_t ninstrs;")
    print("    uint32_t outdegree;")
    print("    const char *priority_class;")
    print("    const char *next_action;")
    print("    const char *selected_range;")
    print("} VMNativeRetdecGapQueueRow;")
    print("")
    print("static const VMNativeRetdecGapQueueRow k_vm_native_retdec_gap_queue[] = {")
    for row in rows:
        print(
            f"    {{ {row['rank']}u, {row['offset']}ull, {row['end']}ull, {row['bytes']}u, "
            f"{row['semantic_covered_bytes']}u, {row['semantic_uncovered_bytes']}u, {row['coverage_x100']}u, "
            f"{cstr(row['kind'])}, {cstr(row['name'])}, {row['nbbs']}u, {row['ninstrs']}u, {row['outdegree']}u, "
            f"{cstr(row['priority_class'])}, {cstr(row['next_action'])}, {cstr(row['selected_range'])} }},"
        )
    print("};")
    print("")
    print("static unsigned vm_native_retdec_gap_queue_count(void) {")
    print("    return (unsigned)(sizeof(k_vm_native_retdec_gap_queue) / sizeof(k_vm_native_retdec_gap_queue[0]));")
    print("}")
    print("")
    print(f"_Static_assert(sizeof(k_vm_native_retdec_gap_queue) / sizeof(k_vm_native_retdec_gap_queue[0]) == {len(rows)}u,")
    print("               \"native RetDec gap queue count changed\");")


def main():
    parser = argparse.ArgumentParser(description="Rank native function skeletons that still lack semantic C coverage.")
    parser.add_argument("--inventory", default=str(TRACE_DIR / "vm_native_function_inventory.tsv"))
    parser.add_argument("--coverage", default=str(TRACE_DIR / "vm_native_executable_coverage_audit.tsv"))
    parser.add_argument("--from-tsv")
    group = parser.add_mutually_exclusive_group()
    group.add_argument("--c", action="store_true")
    group.add_argument("--markdown", action="store_true")
    args = parser.parse_args()

    rows = read_tsv(args.from_tsv) if args.from_tsv else build_rows(args.inventory, args.coverage)
    if args.c:
        emit_c(rows)
    elif args.markdown:
        emit_md(rows)
    else:
        emit_tsv(rows)


if __name__ == "__main__":
    main()

File: dumps/vmtail-wide-1m-w16/vm_native_retdec_gap_queue.c
/*
 * Native RetDec gap queue.
 * Ranks native function skeletons that still lack semantic C sidecar coverage.
 */
#include <stdint.h>

typedef struct VMNativeRetdecGapQueueRow {
    uint32_t rank;
    uint64_t offset;
    uint64_t end;
    uint32_t bytes;
    uint32_t semantic_covered_bytes;
    uint32_t semantic_uncovered_bytes;
    uint32_t coverage_x100;
    const char *kind;
    const char *name;
    uint32_t nbbs;
    uint32_t ninstrs;
    uint32_t outdegree;
    const char *priority_class;
    const char *next_action;
    const char *selected_range;
} VMNativeRetdecGapQueueRow;

static const VMNativeRetdecGapQueueRow k_vm_native_retdec_gap_queue[] = {
    { 1u, 0x6b7a0ull, 0x6b8bbull, 283u, 267u, 16u, 9434u, "r2_discovered", "fcn.0006b7a0", 15u, 78u, 8u, "tier0_small_native_gap", "targeted_retdec_single_function", "0x6b7a0-0x6b8bb" },
    { 2u, 0x57cf40ull, 0x57d4d7ull, 1431u, 0u, 1431u, 0u, "r2_discovered", "fcn.0057cf40", 99u, 399u, 29u, "tier1_medium_native_gap", "targeted_retdec_single_function", "0x57cf40-0x57d4d7" },
    { 3u, 0x4d1510ull, 0x4d19b5ull, 1189u, 0u, 1189u, 0u, "r2_discovered", "fcn.004d1510", 50u, 294u, 15u, "tier1_medium_native_gap", "targeted_retdec_single_function", "0x4d1510-0x4d19b5" },
    { 4u, 0x4b3910ull, 0x4b3cceull, 958u, 0u, 958u, 0u, "r2_discovered", "fcn.004b3910", 40u, 220u, 35u, "tier1_medium_native_gap", "targeted_retdec_single_function", "0x4b3910-0x4b3cce" },
    { 5u, 0x4aa830ull, 0x4aab60ull, 816u, 416u, 400u, 5098u, "r2_discovered", "fcn.004aa830", 32u, 194u, 13u, "tier1_medium_native_gap", "targeted_retdec_single_function", "0x4aa830-0x4aab60" },
    { 6u, 0x415aeull, 0x418c6ull, 792u, 402u, 390u, 5075u, "r2_discovered", "fcn.000415ae", 71u, 235u, 2u, "tier1_medium_native_gap", "targeted_retdec_single_function", "0x415ae-0x418c6" },
    { 7u, 0x554b00ull, 0x554de0ull, 736u, 368u, 368u, 5000u, "r2_discovered", "fcn.00554b00", 37u, 197u, 7u, "tier1_medium_native_gap", "targeted_retdec_single_function", "0x554b00-0x554de0" },
    { 8u, 0x57c770ull, 0x57c987ull, 535u, 416u, 119u, 7775u, "r2_discovered", "fcn.0057c770", 36u, 146u, 5u, "tier1_medium_native_gap", "targeted_retdec_single_function", "0x57c770-0x57c987" },
    { 9u, 0x4926daull, 0x49295cull, 642u, 641u, 1u, 9984u, "r2_discovered", "fcn.004926da", 27u, 153u, 12u, "tier1_medium_native_gap", "targeted_retdec_single_function", "0x4926da-0x49295c" },
    { 10u, 0x4a4be0ull, 0x4a6a5eull, 7806u, 0u, 7806u, 0u, "r2_discovered", "fcn.004a4be0", 324u, 1681u, 95u, "tier2_large_native_gap", "targeted_retdec_with_timeout", "0x4a4be0-0x4a6a5e" },
    { 11u, 0x3e19cull, 0x3f89full, 5891u, 207u, 5684u, 351u, "r2_discovered", "fcn.0003e19c", 468u, 1574u, 1u, "tier2_large_native_gap", "targeted_retdec_with_timeout", "0x3e19c-0x3f89f" },
};

static unsigned vm_native_retdec_gap_queue_count(void) {
    return (unsigned)(sizeof(k_vm_native_retdec_gap_queue) / sizeof(k_vm_native_retdec_gap_queue[0]));
}

_Static_assert(sizeof(k_vm_native_retdec_gap_queue) / sizeof(k_vm_native_retdec_gap_queue[0]) == 11u,
               "native RetDec gap queue count changed");
